### ISSUE TITLE
Image/label values normalisation

### ISSUE THREAD

**Issue by mathpluscode** [DeepRegNet/DeepReg#233]
# Issue description

Sometimes, the data from nifti or from h5 might not have the range as we expected.

For now, we expect:
- Images have values between 0-255
- Labels have values between 0-1

And so we will divide the image tensors by 255.

But sometimes this might not be the case and maybe user would like to define their own value range.
So we can do the following instead,
- For each file loader, h5 or nifti, they will need a value range `[min, max]` and they will normalize the arrays using the given values by `x = (x - min) / (max - min + EPS)` with `EPS=1.0E-6` a very small number. The min and max can be None (any one of them, not necessarily both), and if None, the min/max will be calculated dynamically on the fly. So the min/max would not be global, but be array dependant.
- Then, for image file loader we will assume the max = 255, min = 0, for label file loader we will assume the max = 1, min = 0. However, user can overwrite this values via config. But we do not have to put these values in config, it's optional. Otherwise the config will be too complicated.

What do you think?
@ebonmati @NMontanaBrown @YipengHu 

## Type of Issue

- [ ] New feature request
- [ ] Documentation update
- [ ] Test request

**Comment by YipengHu** [comment]
It is error-prone with anything "we assume". i think by default, it should be x = (x - min) / (max - min + EPS) - calculated per image, unless a global customised [min max] is given; - no need for 255 as a special case.

Another important normalisation is unit-variance with zero-mean. Not sure i see clear advantage, so let's leave it post-r1 when there is a need.

**Comment by YipengHu** [comment]
for labels, just `bool()` the value, if you like to be more thoughtful you can have a customisable threshold - but i think it is an overkill.

**Comment by mathpluscode** [comment]
@YipengHu there's a difference between using a global min/max and using a sample dependent min/max.

Because maybe some image just does not have extreme values like 0 or 255, but others have. I would prefer to make the meaning of a normalized 0.1 fixed, instead of making this dynamic.

So I argue for a default 0 - 255 and user can config otherwise.

For the label, I would not do bool, as It might be a continuous value?

**Comment by YipengHu** [comment]
- with batch norm, the difference between global normalisation and per-image normalisation diminishes;
- besides, what is the reason the global normalisation is better?
- some medical images has default range other than 8 bit, e.g. 16 bits for CT. don't see the reason 255 should be default;
- perhaps more importantly, are you feeding [0, 255] directly into network? if so, you will have two ranges working with DeepReg, [0,255] and [0,1] - this is not ideal as may even more challenging to find good default values for other hyperparameters, e.g. LR; if not (i.e. also rescaling to [0,1]), so what is the point for this default really? 
- good point on "probability-like" labels! need to be very clear on the doc.

**Comment by mathpluscode** [comment]
I divide it by 255 because I don't want to feed it with different ranges. So fine, agree that with batch norm the difference is little. Will change so, and change the configs to make it clear, unless given min/max in the config, we will normalize dynamically.

**Comment by YipengHu** [comment]
> I divide it by 255 because I don't want to feed it with different ranges. So fine, agree that with batch norm the difference is little. Will change so, and change the configs to make it clear, unless given min/max in the config, we will normalize dynamically.

@mathpluscode i see your point - just conventions - pick one and clarify it in config/doc.
